File: go/ontology/registry.go
// SPDX-License-Identifier: Apache-2.0
package ontology

import (
	"context"
	"fmt"
	"strings"
	"time"
)

// ProposeDedupThreshold is the Jaro-Winkler similarity threshold above
// which ProposeType skips registration (considers the type a duplicate).
const ProposeDedupThreshold = 0.85

// Registry provides high-level ontology operations on top of an
// OntologyStore. It adds ProposeType with fuzzy dedup guard,
// RegisterType, DeprecateType, and ApplyTemplate.
type Registry struct {
	store OntologyStore
}

// NewRegistry creates a Registry backed by the given OntologyStore.
func NewRegistry(store OntologyStore) *Registry {
	return &Registry{store: store}
}

// RegisterType adds or updates a type at the specified scope.
func (r *Registry) RegisterType(ctx context.Context, scope Scope, def TypeDefinition) error {
	return r.store.UpsertType(ctx, scope, def)
}

// DeprecateType marks a type as deprecated at the specified scope.
func (r *Registry) DeprecateType(ctx context.Context, scope Scope, typeName string) error {
	existing, err := r.store.GetType(ctx, scope, typeName)
	if err != nil {
		return fmt.Errorf("ontology: deprecate type %q at %s: %w", typeName, scope, err)
	}
	deprecated := *existing
	deprecated.Status = TypeStatusDeprecated
	return r.store.UpsertType(ctx, scope, deprecated)
}

// ProposeType creates a proposed type, skipping if a fuzzy-similar type
// already exists (Jaro-Winkler >= 0.85). Returns nil without error when
// the type is skipped as a duplicate.
func (r *Registry) ProposeType(ctx context.Context, scope Scope, category string, typeName string, reason string) error {
	resolved, err := r.store.GetResolvedOntology(ctx, "", "", "")
	if err != nil {
		return fmt.Errorf("ontology: propose type resolve: %w", err)
	}

	var existingLabels []string
	switch category {
	case "nodeType":
		for _, rt := range resolved.NodeTypes {
			existingLabels = append(existingLabels, rt.Label)
		}
	case "edgeType":
		for _, rt := range resolved.EdgeTypes {
			existingLabels = append(existingLabels, rt.Label)
		}
	default:
		return fmt.Errorf("ontology: category must be nodeType or edgeType, got %q", category)
	}

	proposedLabel := FormatNodeTypeLabel(typeName)
	if category == "edgeType" {
		proposedLabel = FormatEdgeTypeLabel(typeName)
	}

	for _, existing := range existingLabels {
		sim := jaroWinkler(strings.ToLower(proposedLabel), strings.ToLower(existing))
		if sim >= ProposeDedupThreshold {
			return nil
		}
	}

	now := time.Now().UTC().Format(time.RFC3339)
	def := TypeDefinition{
		Type:           typeName,
		Label:          proposedLabel,
		Description:    reason,
		DiscoveredFrom: "proposal",
		CreatedAt:      now,
		Status:         TypeStatusProposed,
	}
	return r.store.UpsertType(ctx, scope, def)
}

// Resolve returns the fully merged ontology from the store.
func (r *Registry) Resolve(ctx context.Context, brainID, projectID, orgID string) (*ResolvedOntology, error) {
	return r.store.GetResolvedOntology(ctx, brainID, projectID, orgID)
}

// jaroWinkler computes Jaro-Winkler similarity between two lowercase
// strings. This is a standalone implementation to avoid depending on
// the dedup package (P6-5).
func jaroWinkler(s1, s2 string) float64 {
	if s1 == s2 {
		return 1.0
	}
	if len(s1) == 0 || len(s2) == 0 {
		return 0.0
	}

	maxDist := len(s1)
	if len(s2) > maxDist {
		maxDist = len(s2)
	}
	matchWindow := maxDist/2 - 1
	if matchWindow < 0 {
		matchWindow = 0
	}

	s1Matches := make([]bool, len(s1))
	s2Matches := make([]bool, len(s2))

	matches := 0
	transpositions := 0

	for i := 0; i < len(s1); i++ {
		start := i - matchWindow
		if start < 0 {
			start = 0
		}
		end := i + matchWindow + 1
		if end > len(s2) {
			end = len(s2)
		}
		for j := start; j < end; j++ {
			if s2Matches[j] || s1[i] != s2[j] {
				continue
			}
			s1Matches[i] = true
			s2Matches[j] = true
			matches++
			break
		}
	}

	if matches == 0 {
		return 0.0
	}

	k := 0
	for i := 0; i < len(s1); i++ {
		if !s1Matches[i] {
			continue
		}
		for !s2Matches[k] {
			k++
		}
		if s1[i] != s2[k] {
			transpositions++
		}
		k++
	}

	jaro := (float64(matches)/float64(len(s1)) +
		float64(matches)/float64(len(s2)) +
		float64(matches-transpositions/2)/float64(matches)) / 3.0

	commonPrefix := 0
	limit := 4
	if len(s1) < limit {
		limit = len(s1)
	}
	if len(s2) < limit {
		limit = len(s2)
	}
	for i := 0; i < limit; i++ {
		if s1[i] != s2[i] {
			break
		}
		commonPrefix++
	}

	return jaro + float64(commonPrefix)*0.1*(1.0-jaro)
}

File: sdks/ts/memory/src/ontology/registry.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * High-level ontology Registry that wraps an OntologyStore and provides
 * ProposeType (with Jaro-Winkler dedup guard >= 0.85), RegisterType,
 * DeprecateType, and Resolve.
 *
 * Port of go/ontology/registry.go.
 */

import type { OntologyScope, OntologyTypeDefinition, TypeStatus } from './types.js'
import type { OntologyStore, ResolvedOntology } from './store.js'
import { formatEdgeTypeLabel, formatNodeTypeLabel } from './format.js'

/**
 * Jaro-Winkler similarity threshold above which ProposeType considers
 * the candidate a duplicate and skips registration.
 */
export const PROPOSE_DEDUP_THRESHOLD = 0.85

export type RegistryOptions = {
  readonly store: OntologyStore
}

/**
 * Registry provides high-level ontology operations on top of an
 * OntologyStore. Adds ProposeType with fuzzy dedup guard, RegisterType,
 * DeprecateType, and Resolve.
 */
export class Registry {
  private readonly store: OntologyStore

  constructor(options: RegistryOptions) {
    this.store = options.store
  }

  /**
   * Adds or updates a type at the specified scope.
   */
  async registerType(scope: OntologyScope, def: OntologyTypeDefinition): Promise<void> {
    await this.store.upsertType(scope, def)
  }

  /**
   * Marks a type as deprecated at the specified scope.
   */
  async deprecateType(scope: OntologyScope, typeName: string): Promise<void> {
    const existing = await this.store.getType(scope, typeName)
    if (existing === undefined) {
      throw new Error(`ontology: type "${typeName}" not found at scope ${scope}`)
    }
    const deprecated: OntologyTypeDefinition = {
      ...existing,
      status: 'deprecated' as TypeStatus,
    }
    await this.store.upsertType(scope, deprecated)
  }

  /**
   * Creates a proposed type, skipping if a fuzzy-similar type already
   * exists (Jaro-Winkler >= 0.85). Returns without error when the type
   * is skipped as a duplicate.
   */
  async proposeType(
    scope: OntologyScope,
    category: 'nodeType' | 'edgeType',
    typeName: string,
    reason: string,
  ): Promise<void> {
    const resolved = await this.store.getResolvedOntology('', '', '')

    const existingLabels: string[] = []
    if (category === 'nodeType') {
      for (const rt of resolved.nodeTypes) {
        existingLabels.push(rt.label)
      }
    } else {
      for (const rt of resolved.edgeTypes) {
        existingLabels.push(rt.label)
      }
    }

    const proposedLabel =
      category === 'nodeType' ? formatNodeTypeLabel(typeName) : formatEdgeTypeLabel(typeName)

    for (const existing of existingLabels) {
      const sim = jaroWinkler(proposedLabel.toLowerCase(), existing.toLowerCase())
      if (sim >= PROPOSE_DEDUP_THRESHOLD) {
        return
      }
    }

    const now = new Date().toISOString()
    const def: OntologyTypeDefinition = {
      type: typeName,
      label: proposedLabel,
      description: reason,
      discoveredFrom: 'proposal',
      createdAt: now,
      status: 'proposed',
    }
    await this.store.upsertType(scope, def)
  }

  /**
   * Returns the fully merged ontology from the store.
   */
  async resolve(brainId: string, projectId: string, orgId: string): Promise<ResolvedOntology> {
    return this.store.getResolvedOntology(brainId, projectId, orgId)
  }
}

/**
 * Standalone Jaro-Winkler similarity for the propose dedup guard.
 * Avoids depending on the dedup package (P6-5).
 */
function jaroWinkler(s1: string, s2: string): number {
  if (s1 === s2) return 1.0
  if (s1.length === 0 || s2.length === 0) return 0.0

  const matchWindow = Math.max(0, Math.floor(Math.max(s1.length, s2.length) / 2) - 1)
  const s1Matches = new Array<boolean>(s1.length).fill(false)
  const s2Matches = new Array<boolean>(s2.length).fill(false)

  let matches = 0
  let transpositions = 0

  for (let i = 0; i < s1.length; i++) {
    const start = Math.max(0, i - matchWindow)
    const end = Math.min(i + matchWindow + 1, s2.length)
    for (let j = start; j < end; j++) {
      if (s2Matches[j] || s1[i] !== s2[j]) continue
      s1Matches[i] = true
      s2Matches[j] = true
      matches++
      break
    }
  }

  if (matches === 0) return 0.0

  let k = 0
  for (let i = 0; i < s1.length; i++) {
    if (!s1Matches[i]) continue
    while (!s2Matches[k]) k++
    if (s1[i] !== s2[k]) transpositions++
    k++
  }

  const jaro =
    (matches / s1.length + matches / s2.length + (matches - transpositions / 2) / matches) / 3

  let commonPrefix = 0
  const maxPrefix = Math.min(4, Math.min(s1.length, s2.length))
  for (let i = 0; i < maxPrefix; i++) {
    if (s1[i] !== s2[i]) break
    commonPrefix++
  }

  return jaro + commonPrefix * 0.1 * (1 - jaro)
}
